woocommerce/woocommerce-ios
# Store creation: update how we wait for the new site to be ready for the timeout issue

Part of #9733 


## Why

Since May 4, we started seeing most created sites timing out while waiting for the site to be a Jetpack/Woo site (more details in pe5sF9-1vz-p2). This means that most of the merchants see a timeout alert in the app, instead of the success state where they're navigated to the newly created site. Based on Dale@WPCOM's advice pe5sF9-1vz-p2#comment-2166, we can update how we wait for the site from polling from `me/sites` to `sites/:siteId/plugins` (to check if Jetpack-the-plugin is active) & `sites/:siteId` (to fetch site name etc.) as a workaround.

## How

In iOS, we currently call `me/sites` periodically to check if the site is ready or not. We're updating this endpoint to `sites/:siteId/plugins` to wait for Jetpack-the-plugin to be active. When it's active, we then call `sites/:siteId` to sync the basic site info like the site name. Some properties of the site might still be out of sync, and we should log the error `site_creation_properties_out_of_sync` like before. The API call sequence looks like:

<img width="1077" alt="Screenshot 2023-05-16 at 2 35 40 PM" src="https://github.com/woocommerce/woocommerce-ios/assets/1945542/8950f6ee-c6e0-49be-ae6e-fb8df19d509e">

After a few failed API calls to `sites/:siteId/plugins`, it succeeds eventually and then `sites/:siteId` is called to fetch the site info (which might still be out-of-sync due to the API limitation). Please note that `sites/:siteId/plugins` is called again after `sites/:siteId` because the retry still checks for the plugins endpoint after throwing an error for logging. I plan to refactor this part of the code to remove this extra call and to be testable in a separate future task of #9733.

## Changes

Two endpoints were added to the Networking layer:

- `sites/:siteId/plugins`: `SitePluginsRemote.loadPluginsFromWPCOM` loads all the site plugins from WPCOM (to differentiate from `SitePluginsRemote.loadPlugins` which loads all the site plugins from the site)
- `sites/:siteId`: `SiteRemote.loadSite` loads a single site given the site ID

Then these two endpoints are used in two new Yosemite actions:

- `SitePluginAction.isPluginActive`: this calls `SitePluginsRemote.loadPluginsFromWPCOM` to load all the site plugins from WPCOM and then checks if a plugin is active
- `SiteAction.syncSite`: this calls `SiteRemote.loadSite` to load a single site from WPCOM and then upserts to storage

In the store creation flow where we wait for the site to be ready in `StoreCreationCoordinator.waitForSiteToBecomeJetpackSite`, it first checks if Jetpack-the-plugin is active using `SitePluginAction.isPluginActive`. When the plugin becomes active, it then syncs the site info using `SiteAction.syncSite`.

## Testing instructions


- Log in with a WPCOM account if needed
- Go to the Menu tab, and tap on the store picker
- At the end of the store list, tap `Add a store`
- Continue the steps to create a free trial site --> after about 30-40 seconds, the flow should succeed and navigate to the newly created store without a Jetpack upsell banner (before this PR, the app almost always times out with an alert and never enters the store)


---
